library/ssl_misc.h

```diff
@@ -987,6 +982,11 @@ struct mbedtls_ssl_handshake_params
     } tls13_master_secrets;
 
     mbedtls_ssl_tls13_handshake_secrets tls13_hs_secrets;
+#if defined(MBEDTLS_SSL_EARLY_DATA)
+    mbedtls_ssl_tls13_early_secrets tls13_early_secrets;
+    /** TLS 1.3 transform for 0-RTT application and handshake messages. */
```

Comment:
Will this only used to protect the early data, which is 0-RTT application data? 
Should we put the description of **handshake messages** here?

Reply:
Both are needed. It is not only for early application message but also handshake message like HRR and EOED.

Reply:
Sorry, I made a mistake. It is 2nd Client Hello not HRR. HRR is outbound message .  at this moment, I think it should be `handshake message`

Section A.2 and 2.3 said early data can be sent after ClientHello, and it does not say it MUST be reset before 2nd ClientHello. And the 2nd ClientHello from GnuTLS is protected by early transform. I am not sure if it is right because 

https://github.com/Mbed-TLS/mbedtls/pull/6485#issue-1423357463 , I post the detail  .

Reply:
> Section A.2 and 2.3 said early data can be sent after ClientHello, and it does not say it MUST be reset before 2nd ClientHello. And the 2nd ClientHello from GnuTLS is protected by early transform. I am not sure if it is right because
>

That's a good point. Looking at the spec I am not sure either if when the client started sending early data and receives an HRR it should send the second client hello in plaintext or encrypted. Would you mind asking Hannes about that ?

Reply:
> That's a good point. Looking at the spec I am not sure either if when the client started sending early data and receives an HRR it should send the second client hello in plaintext or encrypted. Would you mind asking Hannes about that ?

Okay. I will ask him.
For time being, I suppose 2nd ClientHello is encrypted. That will work with GnuTLS.

Reply:
Hannes has answered the issue in #6485